# Машинное обучение, ФЭН ВШЭ

## Практическое задание 1

### Общая информация
Дата выдачи: 19.09.2019  

Мягкий дедлайн: 23:59MSK 27.09.2019 (за каждый день просрочки снимается 1 балл)

Жесткий дедлайн: 23:59MSK 04.10.2019  

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В каждом разделе вам предлагается выполнить несколько заданий. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотекам pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после указанного срока сдачи нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

### Формат сдачи
Будет объявлен позже.

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написаем production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [ ]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

А это ___ячейка с текстом___.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

In [1]:
# вот код: 
name = input("Enter your name: ")
print(name, ", ", "you are cute!", sep="")

Enter your name: Natalia
Natalia, you are cute!


[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. #Делать 
##заголовки 
###разного уровня
3. Выделять *текст* <s>при</s> **необходимости**
4. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

А ещё можно вставлять картинки:
<img src="https://st2.depositphotos.com/1177973/9266/i/950/depositphotos_92668716-stock-photo-red-cat-with-computer-keyboard.jpg" style="width: 400px">

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных. С этой библиотекой мы сейчас и познакомимся.

Основной объект в pandas это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [2]:
%pylab inline  
# import almost all we need
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


Данные можно скачать [отсюда](https://www.dropbox.com/s/5qq94wzmbw4e54r/data.csv?dl=0).

#### 1. [0.5 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

In [3]:
data = pd.read_csv("data.csv")
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

*Данные о заказах. 5 столбцов: номер заказа, количество тотваров в заказе, название товара, описание товара, цена.*


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null object
dtypes: int64(2), object(3)
memory usage: 126.4+ KB


*Всего 4622 строчки. 2 числовые переменные: номер заказа и количество товаров в заказе. 3 текстовые переменные: описание товара, название товара и цена.*

#### 2. [0.25 баллов] Ответьте на вопросы:
1. Сколько заказов попало в выборку?
2. Сколько уникальных категорий товара было куплено? (item_name)

In [69]:
len(data.groupby("order_id"))

1834

*В выборку попало 1834 заказа.*

In [70]:
len(data.groupby("item_name"))

50

*Куплено 50 уникальных категорий товаров.*

#### 3. [0.25 баллов] Есть ли в данных пропуски? В каких колонках? 

In [6]:
data.isnull().sum()

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

*В данных есть пропуски. В колонке  choice_description не хватает 1246 строчек данных.*


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [8]:
data =  data.replace(pd.np.nan, '', regex=True)

In [9]:
data.isnull().sum()

order_id              0
quantity              0
item_name             0
choice_description    0
item_price            0
dtype: int64

*Вот, теперь пропущенных данных нет.*

#### 4. [0.5 баллов] Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [10]:
data.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

*Колонка с ценой товара имеет тип object. Строковые данные.*

In [11]:
# Уберу значок доллара.
def get_price(price):
    return price[1:]

prices = data['item_price'].apply(get_price)
prices.head()

0     2.39 
1     3.39 
2     3.39 
3     2.39 
4    16.98 
Name: item_price, dtype: object

In [13]:
data["prices"] = data["prices"].apply(pd.to_numeric)

In [14]:
data.dtypes

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price             object
prices                float64
dtype: object

Какая средняя/минимальная/максимальная цена у товара? 

In [15]:
# Нахождение минимальной цены
data["prices"].min()

1.09

In [16]:
# Нахождение максимальной цены
data["prices"].max()

44.25

In [17]:
# Нахождение средней цены
data["prices"].mean()

7.464335785374397

Удалите старую колонку с ценой.

In [18]:
data.drop("item_price", axis = 1, inplace=True)
data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  prices  
0                                                       2.39  
1                                       [Clementine]    3.39  
2                                            [Apple]    3.39  
3                                                       2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   16.98

#### 5. [0.25 баллов] Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по item_name)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

In [27]:
data1 = data.drop_duplicates(subset ="item_name", keep = "first", inplace=False)

In [28]:
len(data1)

50

*Избавилась от дубликатов. Осталось 50 уникальных товаров, как и было получено в первых заданиях.*

In [29]:
data1.sort_values(by=['prices'], ascending=True).head(5)

order_id  quantity          item_name choice_description  prices
34         17         1      Bottled Water                       1.09
263       114         1  Canned Soft Drink             [Coke]    1.25
6           3         1      Side of Chips                       1.69
40         19         1              Chips                       2.15
18          9         2        Canned Soda           [Sprite]    2.18

*Всё плохо с количеством. Надо, чтобы была указана цена за одну единицу товара.*

In [31]:
data1 = data1.assign(price=data1["prices"]/data1["quantity"])
data1

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
6            3         1                          Side of Chips   
7            4         1                          Steak Burrito   
8            4         1                       Steak Soft Tacos   
10           5         1                    Chips and Guacamole   
11           6         1                   Chicken Crispy Tacos   
12           6         1                     Chicken Soft Tacos   
16           8         1                        Chicken Burrito   
18           9         2                            Canned Soda   
21          11         1                       Barbacoa Burrito   
27          14         1                       Carnitas Burrito   
33          17         1                          Carnitas Bowl   
34          17         1                          Bottled Water   
38          18         1  Chips and Tomatillo Green Chili Salsa   
39          19         1                          Barbacoa Bowl   
40          19         1                                  Chips   
44          20         1                     Chicken Salad Bowl   
54          25         1                             Steak Bowl   
56          26         1                    Barbacoa Soft Tacos   
57          26         1                         Veggie Burrito   
62          28         1                            Veggie Bowl   
92          40         1                     Steak Crispy Tacos   
111         49         1    Chips and Tomatillo Red Chili Salsa   
168         75         1                  Barbacoa Crispy Tacos   
186         83         1                      Veggie Salad Bowl   
191         85         1     Chips and Roasted Chili-Corn Salsa   
233        102         1     Chips and Roasted Chili Corn Salsa   
237        103         1                    Carnitas Soft Tacos   
250        109         1                          Chicken Salad   
263        114         1                      Canned Soft Drink   
281        123         2                       Steak Salad Bowl   
298        129         1                      6 Pack Soft Drink   
300        130         1    Chips and Tomatillo-Red Chili Salsa   
445        193         3                                   Bowl   
510        214         1                                Burrito   
520        217         1                           Crispy Tacos   
554        230         1                  Carnitas Crispy Tacos   
664        276         1                            Steak Salad   
674        279         1      Chips and Mild Fresh Tomato Salsa   
738        304         1                      Veggie Soft Tacos   
1132       468         1                    Carnitas Salad Bowl   
1229       501         1                    Barbacoa Salad Bowl   
1414       575         1                                  Salad   
1653       668         1                    Veggie Crispy Tacos   
1694       686         1                           Veggie Salad   
3750      1500         1                         Carnitas Salad   

                                     choice_description  prices  price  
0                                                          2.39   2.39  
1                                          [Clementine]    3.39   3.39  
2                                               [Apple]    3.39   3.39  
3                                                          2.39   2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   16.98   8.49  
6                                                          1.69   1.69  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...   11.75  11.75  
8   

*Вот топ-5 самых дешёвых товаров.*

In [35]:
data1.sort_values(by=['price']).head(5)

order_id  quantity          item_name choice_description  prices  price
34         17         1      Bottled Water                       1.09   1.09
18          9         2        Canned Soda           [Sprite]    2.18   1.09
263       114         1  Canned Soft Drink             [Coke]    1.25   1.25
6           3         1      Side of Chips                       1.69   1.69
40         19         1              Chips                       2.15   2.15

*Вот топ-5 самых дорогих товаров.*

In [36]:
data1.sort_values(by=['price']).tail(5)

order_id  quantity              item_name  \
7            4         1          Steak Burrito   
168         75         1  Barbacoa Crispy Tacos   
1132       468         1    Carnitas Salad Bowl   
1229       501         1    Barbacoa Salad Bowl   
281        123         2       Steak Salad Bowl   

                                     choice_description  prices  price  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...   11.75  11.75  
168   [Tomatillo Red Chili Salsa, [Rice, Black Beans...   11.75  11.75  
1132  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...   11.89  11.89  
1229  [Fresh Tomato Salsa, [Rice, Fajita Vegetables,...   11.89  11.89  
281   [Tomatillo Red Chili Salsa, [Black Beans, Chee...   23.78  11.89

#### 6. [0.5 баллов] Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [39]:
data2 = data.groupby('order_id')["quantity"].sum()
data2.mean()

2.711014176663032

*Вот сколько в среднем покупают товаров за один заказ.*

*Теперь можно узнать средную стоимость одного заказа.*

In [40]:
data3 = data.groupby('order_id')["prices"].sum()
data3.mean()

18.81142857142869

#### 7. [0.25 баллов] Сколько заказов содержали ровно 1 товар?

In [43]:
data4 = data.groupby("order_id").sum()
data4 = data4[data4["quantity"] == 1].count()
data4

quantity    56
prices      56
dtype: int64

*56 заказов содержат ровно 1 товар.*

#### 8. [0.75 баллов] Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [44]:
ndata = np.random.rand(10, 3)
ndata

array([[0.36970192, 0.39405675, 0.36855143],
       [0.9913297 , 0.63789045, 0.54256837],
       [0.56982092, 0.45368343, 0.94976565],
       [0.79572221, 0.87258216, 0.06402426],
       [0.22672408, 0.13381498, 0.60651368],
       [0.89668951, 0.11126096, 0.12931234],
       [0.78210302, 0.76636854, 0.96268126],
       [0.66726052, 0.30771047, 0.83337417],
       [0.90304656, 0.51153666, 0.84509389],
       [0.30032692, 0.93393096, 0.23087797]])

In [45]:
df = pd.DataFrame(ndata, columns=["index", "column1", "column2"])
df

index   column1   column2
0  0.369702  0.394057  0.368551
1  0.991330  0.637890  0.542568
2  0.569821  0.453683  0.949766
3  0.795722  0.872582  0.064024
4  0.226724  0.133815  0.606514
5  0.896690  0.111261  0.129312
6  0.782103  0.766369  0.962681
7  0.667261  0.307710  0.833374
8  0.903047  0.511537  0.845094
9  0.300327  0.933931  0.230878

In [46]:
df = df.set_index(df.columns[0])
df

column1   column2
index                       
0.369702  0.394057  0.368551
0.991330  0.637890  0.542568
0.569821  0.453683  0.949766
0.795722  0.872582  0.064024
0.226724  0.133815  0.606514
0.896690  0.111261  0.129312
0.782103  0.766369  0.962681
0.667261  0.307710  0.833374
0.903047  0.511537  0.845094
0.300327  0.933931  0.230878

Сохраните DataFrame на диск в формате csv без индексов и названий столбцов.

In [47]:
df.to_csv('ndata.csv', index=False)

## 2. Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. В этом разделе мы этим и займёмся.

У matplotlib, конечно, же есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторое величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить каждый из них. Не менее важно усвоить базовые принципы визуализаций:
- на графиках должны быть подписаны оси;
- у визуализации должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах).
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать.

In [ ]:
%matplotlib inline  # нужно для отображения графиков внутри ноутбука
import matplotlib.pyplot as plt

На самом деле мы уже импортировали matplotlib внутри %pylab inline в начале задания.

Работать мы будем с той же выборкой покупкок. Добавим новую колонку с датой покупки.

In [ ]:
import datetime

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pd.DataFrame(index=df.order_id.unique())
dates['date'] = [
    (start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
    for _ in range(df.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
df['date'] = df.order_id.map(dates['date'])

#### 1. [1 балл] Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов item_name. 

Изображайте на двух соседних графиках. Для этого может быть полезен subplot.

In [ ]:
# your code

#### 2. [1 балл] Постройте график зависимости суммы покупок от дней.

In [ ]:
# your code

#### 3. [1 балл] Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
# your code

#### 4. [1 балл] Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
# your code

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# your code

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё. 